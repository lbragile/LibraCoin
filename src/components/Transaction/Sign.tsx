import React, { useContext } from "react";

import { Form, Button, InputGroup } from "react-bootstrap";
import { AppContext } from "../../context/AppContext";
import { IState } from "../../typings/AppTypes";

export interface ISign {
  validated: boolean;
  signed: boolean;
  handleSubmit: (e: React.FormEvent<HTMLFormElement>) => void;
}

export default function Sign({ validated, signed, handleSubmit }: ISign): JSX.Element {
  const { state } = useContext(AppContext) as { state: IState };

  function checkAmount(e: React.FocusEvent<HTMLInputElement>): void {
    const userBalance = state.user.balance ?? 1000;
    e.target.value = Math.min(Math.max(0.1, +e.target.value), userBalance).toFixed(2);
  }

  return (
    <Form noValidate validated={validated} className="col-12 col-lg-5 trans-form" onSubmit={handleSubmit}>
      <Form.Group>
        <Form.Control className="text-truncate" type="text" defaultValue={state.user.publicKey ?? ""} disabled={true} />
        <Form.Text className="text-muted">
          Your public key → used to verify transaction was signed using your private key
        </Form.Text>
        <h3 className="my-0 text-center">↓</h3>
        <Form.Control
          className="text-truncate"
          type="text"
          placeholder="Receiver's public key"
          pattern="[A-Za-z0-9]{182,182}"
          required
        />
        <Form.Control.Feedback type="invalid">
          <b>Length or format are incorrect!</b>
        </Form.Control.Feedback>
      </Form.Group>

      <Form.Group>
        <InputGroup className="mb-2">
          <Form.Control
            type="number"
            step="any"
            placeholder={Number(1).toFixed(2)}
            onBlur={(e: React.FocusEvent<HTMLInputElement>) => checkAmount(e)}
            required
          />
          <InputGroup.Prepend>
            <InputGroup.Text>LC</InputGroup.Text>
          </InputGroup.Prepend>
        </InputGroup>
      </Form.Group>

      <Form.Group>
        <Form.Control as="textarea" rows={2} placeholder="optional message..." />
      </Form.Group>

      <Form.Group>
        <Form.Control
          className="text-truncate"
          type="text"
          defaultValue={state.user.privateKey ?? ""}
          disabled={true}
        />
        <Form.Text className="text-muted">Your private key → not shared with anyone, keep this secret!</Form.Text>
      </Form.Group>

      <Button variant="primary" type="submit" disabled={signed} block>
        <b>Sign</b>
      </Button>
    </Form>
  );
}
