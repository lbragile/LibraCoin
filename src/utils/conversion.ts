import { TextEncoder } from "text-encoding";

export function bufferToHex(hashBuffer: ArrayBuffer): string {
  const hashArray = Array.from(new Uint8Array(hashBuffer)); // convert buffer to byte array
  const hashHex = hashArray.map((b) => b.toString(16).padStart(2, "0")).join(""); // convert bytes to hex string
  return hashHex;
}

export async function CryptoKeyToHex(format: string, key: CryptoKey): Promise<string> {
  const buf = (await window.crypto.subtle.exportKey(format, key)) as ArrayBuffer;
  return bufferToHex(buf);
}

export async function digestMessage(message: string): Promise<string> {
  const msgUint8 = new TextEncoder().encode(message); // encode as (utf-8) Uint8Array
  const hashBuffer = await window.crypto.subtle.digest("SHA-256", msgUint8); // hash the message
  return bufferToHex(hashBuffer);
}

export function randomHash(len: number): string {
  return bufferToHex(window.crypto.getRandomValues(new Uint32Array(len)));
}
