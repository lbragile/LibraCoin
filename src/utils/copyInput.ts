export function copyKey(
  e: React.FocusEvent<HTMLTextAreaElement | HTMLInputElement>,
  setCopied: (arg: boolean[]) => void,
  type?: "public" | "private"
): void {
  e.target.select();
  e.target.setSelectionRange(0, 1e6);
  document.execCommand("copy");

  // wallet page, copying keys
  if (type) {
    const isPublic = type === "public";
    setCopied([isPublic, !isPublic && !e.target.value.includes("â—¦")]);
  } else {
    setCopied([true]);
  }
}
