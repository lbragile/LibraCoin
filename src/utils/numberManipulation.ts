// https://stackoverflow.com/a/2901298/4298115
export function numberWithCommas(x: number): string {
  return x.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",");
}
