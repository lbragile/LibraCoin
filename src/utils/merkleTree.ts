import { Tree } from "./Tree";
import { ITransaction } from "../typings/AppTypes";
import { digestMessage } from "./conversion";

export async function calculateMerkleTreeFormation(
  verifiedTrans: ITransaction[],
  selectedTrans: ITransaction[],
  setMerkleTree?: (arg: string[][]) => void
): Promise<string> {
  let tree = [[""]];
  if (selectedTrans.length > 0) {
    // need to make sure node's in tree appear in same order as in the verified transaction pane, regardless of selection order
    const verifiedSignatures = verifiedTrans.map((trans) => trans.signature);
    const selectedSignatures = selectedTrans.map((trans) => trans.signature);
    let signatures = verifiedSignatures.filter((sig) => selectedSignatures.includes(sig));
    tree = [signatures];

    while (signatures.length !== 1) {
      const hashArr = [] as string[];
      for (let i = 0; i < signatures.length; i += 2) {
        const hash = signatures[i + 1] ? await digestMessage(signatures[i] + signatures[i + 1]) : signatures[i];
        hashArr.push(hash);
      }

      signatures = hashArr;
      tree.push(hashArr);
    }
  }

  if (setMerkleTree) {
    setMerkleTree(tree);
  }

  return getMerkleRoot(tree);
}

export function getMerkleRoot(tree: string[][]): string {
  return tree[tree.length - 1][0];
}

export function flattenTree(tree: string[][]): string[] {
  let flatTree = [] as string[];
  // need to reverse each tree row so that printing produces the right order!
  for (let i = 0; i < tree.length; i++) {
    flatTree = flatTree.concat(tree[i].reverse());
  }

  return flatTree;
}

export function drawTreeDiagramOnCanvas(
  merkleTree: string[][],
  canvas: HTMLCanvasElement | null,
  transactions: ITransaction[]
): void {
  if (canvas) {
    const canvasTree = new Tree(canvas, transactions);
    canvasTree.clear();
    const flatTree = flattenTree(merkleTree);
    for (let i = flatTree.length - 1; i >= 0; i--) {
      canvasTree.addNode(flatTree[i]);
    }
    canvasTree.drawTree();
  }
}
