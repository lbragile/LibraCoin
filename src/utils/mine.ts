import { digestMessage, randomHash } from "./conversion";

async function createTarget(numZeros: number): Promise<string> {
  const targetHash = await digestMessage(randomHash(20));

  // replace leading bits with zeros
  const re = new RegExp(`^.{0,${numZeros}}`, "g");
  const zerosStr = Array(numZeros).fill("0").join("");

  return targetHash.replace(re, zerosStr);
}

export async function mine(
  nonce: number,
  setHeader: (arg: number) => void,
  setTarget: (arg: string) => void,
  setSolution: (arg: string) => void,
  setIsValid: (arg: boolean) => void
): Promise<void> {
  // make target
  const numZeros = Math.round(Math.random()) + 2;
  const targetHash = await createTarget(numZeros);
  setTarget(targetHash);

  // mine
  let candidateSolution = "";
  let header = nonce; // re-assigning for clarity
  while (header <= Number.MAX_SAFE_INTEGER) {
    candidateSolution = await digestMessage(header.toString());
    setSolution(candidateSolution);
    setHeader(header++);

    const leadingBits = candidateSolution.substr(0, numZeros).split("");
    if (leadingBits.every((bit) => bit === "0")) {
      break;
    }
  }

  setIsValid(candidateSolution <= targetHash);
}
