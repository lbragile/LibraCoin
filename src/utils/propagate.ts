import { ACTIONS } from "../enums/AppDispatchActions";
import { IAction, IBlock, IState, ITransaction } from "../typings/AppTypes";
import { digestMessage } from "./conversion";

export async function propagateBlockStatus(
  state: IState,
  dispatch: React.Dispatch<IAction>,
  index: number,
  prevHash: string,
  currHash: string,
  skipFirstUpdate: boolean,
  newRoot?: string,
  transactions?: ITransaction[]
): Promise<void> {
  const newBlocks: IBlock[] = [];
  for (let i = index; i < state.chain.length; i++) {
    const merkleRoot = newRoot && i === index ? newRoot : state.chain[i].merkleRoot;
    const valid = skipFirstUpdate && i === index;
    const showTrans = state.chain[i].showTrans ?? false;
    currHash = i === index ? currHash : await digestMessage(i + prevHash + merkleRoot);
    transactions = i === index && transactions ? transactions : state.chain[i].transactions;

    newBlocks.push({ index: i, timestamp: Date.now(), prevHash, currHash, transactions, merkleRoot, valid, showTrans });

    prevHash = currHash; // next block's prevHash is this block's currHash
  }

  dispatch({ type: ACTIONS.UPDATE_BLOCK, payload: { block: newBlocks } });
}
