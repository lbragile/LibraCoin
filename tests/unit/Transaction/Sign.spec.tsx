/**
 * @group unit
 */

import React from "react";
import { screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import Sign from "../../../src/components/Transaction/Sign";
import * as conversionUtils from "../../../src/utils/conversion";
import { customRender } from "../../utils/testUtils";
import { numberWithCommas } from "../../../src/utils/numberManipulation";

const { initialState } = global;

it("renders correctly", () => {
  const { asFragment } = customRender(<Sign />);

  expect(screen.getByRole("form", { name: /Sign Form/i })).toHaveFormValues({
    from: initialState.user.publicKey,
    to: "",
    amount: null,
    msg: "",
    fromSK: initialState.user.privateKey
  });

  const senderPublicKey = screen.getByRole("textbox", { name: /Sender Public Key/i });
  expect(senderPublicKey).toHaveAttribute("readOnly");
  expect(senderPublicKey).not.toBeRequired();

  expect(screen.getByText(/Used to verify/i)).toBeInTheDocument();

  const receiverPublicKey = screen.getByRole("textbox", { name: /Receiver PK/i });
  expect(receiverPublicKey).toBeEnabled();
  expect(receiverPublicKey).toBeRequired();

  const signAmount = screen.getByRole("spinbutton", { name: /Sign Amount/i });
  expect(signAmount).toBeEnabled();
  expect(signAmount).toHaveAttribute("placeholder", "1.00");
  expect(signAmount).toBeRequired();

  const signMessage = screen.getByRole("textbox", { name: /Sign Message/i });
  expect(signMessage).toBeEnabled();
  expect(signMessage).toHaveAttribute("placeholder", "optional message...");
  expect(signMessage).not.toBeRequired();

  const senderPrivateKey = screen.getByRole("textbox", { name: /Sender Private Key/i });
  expect(senderPrivateKey).toHaveAttribute("readOnly");
  expect(senderPrivateKey).not.toBeRequired();

  expect(screen.getByText(/Not shared with/i)).toBeInTheDocument();

  expect(asFragment()).toMatchSnapshot();
});

describe("sign button state", () => {
  it("is enabled when not signed", () => {
    customRender(<Sign />);

    // fill out the form
    userEvent.type(
      screen.getByRole("textbox", { name: /Receiver PK/i }),
      new Array(initialState.user.publicKey.length).fill("A").join("")
    );

    userEvent.type(screen.getByRole("spinbutton", { name: /Sign Amount/i }), Number(12.31).toFixed(2));

    expect(screen.getByRole("button", { name: /Sign Button/i })).toBeEnabled();
    expect(screen.getByRole("button", { name: /Sign Button/i })).toHaveTextContent("Sign");
    expect(screen.getByRole("button", { name: /Sign Button/i })).toHaveClass("btn-primary");
  });

  it("is disabled when signed", async () => {
    jest.spyOn(conversionUtils, "digestMessage").mockResolvedValueOnce("random digest");
    customRender(<Sign />);

    // fill out the form
    userEvent.type(
      screen.getByRole("textbox", { name: /Receiver PK/i }),
      new Array(initialState.user.publicKey.length).fill("A").join("")
    );
    userEvent.type(screen.getByRole("spinbutton", { name: /Sign Amount/i }), Number(320.12).toFixed(2));
    userEvent.type(screen.getByRole("textbox", { name: /Sign Message/i }), "random{space}message");

    userEvent.click(screen.getByRole("button", { name: /Sign Button/i }));

    await waitFor(() => expect(screen.getByRole("button", { name: /Sign Button/i })).toBeDisabled());
    expect(screen.getByRole("button", { name: /Sign Button/i })).toHaveTextContent("Signed");
    expect(screen.getByRole("button", { name: /Sign Button/i })).toHaveClass("btn-success");
  });
});

describe("Amount checking", () => {
  beforeEach(() => customRender(<Sign />));

  it("shows error when empty", async () => {
    const signAmount = screen.getByRole("spinbutton", { name: /Sign Amount/i });
    signAmount.focus();
    signAmount.blur(); // cause an error

    const alert = await screen.findByRole("alert", { name: /Amount Feedback/i });
    expect(alert).toBeInTheDocument();
    expect(alert).toHaveTextContent("Amount is required!");
  });

  it("shows error when negative", async () => {
    const signAmount = screen.getByRole("spinbutton", { name: /Sign Amount/i });
    userEvent.type(signAmount, "-0.1");
    signAmount.blur(); // cause an error

    const alert = await screen.findByRole("alert", { name: /Amount Feedback/i });
    expect(alert).toBeInTheDocument();
    expect(alert).toHaveTextContent("Cannot be a negative value");
  });

  it("shows error when below minimum (0.10)", async () => {
    const signAmount = screen.getByRole("spinbutton", { name: /Sign Amount/i });
    userEvent.type(signAmount, "0.09");
    signAmount.blur(); // cause an error

    const alert = await screen.findByRole("alert", { name: /Amount Feedback/i });
    expect(alert).toBeInTheDocument();
    expect(alert).toHaveTextContent("Must be at least $0.10 LC");
  });

  it("shows error when above maximum", async () => {
    const signAmount = screen.getByRole("spinbutton", { name: /Sign Amount/i });
    userEvent.type(signAmount, (initialState.user.balance + 0.01).toString());
    signAmount.blur(); // cause an error

    const alert = await screen.findByRole("alert", { name: /Amount Feedback/i });
    expect(alert).toBeInTheDocument();
    expect(alert).toHaveTextContent(`Must be at most $${numberWithCommas(initialState.user.balance)} LC`);
  });

  it("does not show error when between", () => {
    const signAmount = screen.getByRole("spinbutton", { name: /Sign Amount/i });
    userEvent.type(signAmount, "123.4");
    signAmount.blur();

    expect(screen.queryByRole("alert", { name: /Amount Feedback/i })).not.toBeInTheDocument();
  });

  it("shows error when more than 2 decimal places", async () => {
    const signAmount = screen.getByRole("spinbutton", { name: /Sign Amount/i });
    userEvent.type(signAmount, "123.456");
    signAmount.blur(); // cause an error

    const alert = await screen.findByRole("alert", { name: /Amount Feedback/i });
    expect(alert).toBeInTheDocument();
    expect(alert).toHaveTextContent("Allowed at most 2 decimal places");
  });

  it("does not show error for 2 decimal places", () => {
    const signAmount = screen.getByRole("spinbutton", { name: /Sign Amount/i });
    userEvent.type(signAmount, "123.45");
    signAmount.blur();

    expect(screen.queryByRole("alert", { name: /Amount Feedback/i })).not.toBeInTheDocument();
  });
});

describe("form validation", () => {
  test("invalid receiver public key length - empty", async () => {
    customRender(<Sign />);

    const input = screen.getByRole("textbox", { name: /Receiver PK/i });
    input.focus();
    input.blur(); // cause an error

    const alert = await screen.findByRole("alert", { name: /Receiver PK Feedback/i });
    expect(alert).toBeInTheDocument();
    expect(alert).toHaveTextContent("Receiver Public Key is required!");
  });

  test("invalid receiver public key length - short", async () => {
    customRender(<Sign />);

    const input = screen.getByRole("textbox", { name: /Receiver PK/i });
    const text = new Array(initialState.user.publicKey.length - 1).fill("A").join("");
    userEvent.type(input, text);
    input.blur(); // cause an error
    expect(input).toHaveValue(text);

    const alert = await screen.findByRole("alert", { name: /Receiver PK Feedback/i });
    expect(alert).toBeInTheDocument();
    expect(alert).toHaveTextContent("Length is too short");
  });

  test("invalid receiver public key length - long", async () => {
    customRender(<Sign />);

    const input = screen.getByRole("textbox", { name: /Receiver PK/i });
    const text = new Array(initialState.user.publicKey.length + 1).fill("A").join("");
    userEvent.type(input, text);
    input.blur(); // cause an error

    expect(input).toHaveValue(text);

    const alert = await screen.findByRole("alert", { name: /Receiver PK Feedback/i });
    expect(alert).toBeInTheDocument();
    expect(alert).toHaveTextContent("Length is too long");
  });

  test("invalid receiver public key length - alphanumeric", async () => {
    customRender(<Sign />);

    const input = screen.getByRole("textbox", { name: /Receiver PK/i });
    const text = new Array(initialState.user.publicKey.length - 1).fill("A").join("");
    userEvent.type(input, text + "@");
    input.blur(); // cause an error

    expect(input).toHaveValue(text + "@");

    const alert = await screen.findByRole("alert", { name: /Receiver PK Feedback/i });
    expect(alert).toBeInTheDocument();
    expect(alert).toHaveTextContent("Format is invalid, characters must be alphanumeric");
  });

  test("valid receiver public key length", async () => {
    customRender(<Sign />);

    const input = screen.getByRole("textbox", { name: /Receiver PK/i });
    const text = new Array(initialState.user.publicKey.length).fill("A").join("");
    userEvent.type(input, text);
    input.blur(); // cause an error

    expect(input).toHaveValue(text);

    expect(screen.queryByRole("alert", { name: /Receiver PK Feedback/i })).not.toBeInTheDocument();
  });
});
