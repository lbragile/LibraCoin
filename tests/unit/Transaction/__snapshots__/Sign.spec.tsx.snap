// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  <form
    aria-label="Sign Form"
    class="col-12 col-lg-5 trans-form"
    novalidate=""
  >
    <div
      class="input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Sender Public Key
        </span>
      </div>
      <input
        aria-label="Sender Public Key"
        class="text-truncate rounded-right form-control"
        name="from"
        readonly=""
        type="text"
        value="3059301306072a8648ce3d020106082a8648ce3d030107034200048fa7d69599babb"
      />
    </div>
    <small
      class="text-muted form-text"
    >
      Used to verify transaction was signed using your private key
    </small>
    <div
      class="my-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Receiver Public Key
        </span>
      </div>
      <input
        aria-label="Receiver PK"
        class="text-truncate rounded-right form-control"
        name="to"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="mb-2 input-group"
    >
      <input
        aria-label="Sign Amount"
        class="form-control"
        name="amount"
        placeholder="1.00"
        required=""
        step="any"
        type="number"
        value=""
      />
      <div
        class="input-group-append"
      >
        <span
          class="rounded-right border-left-0 input-group-text"
        >
          LC
        </span>
      </div>
    </div>
    <div
      class="mb-2 input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Message
        </span>
      </div>
      <textarea
        aria-label="Sign Message"
        class="form-control"
        name="msg"
        placeholder="optional message..."
        rows="2"
      />
    </div>
    <div
      class="input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Sender Private Key
        </span>
      </div>
      <input
        aria-label="Sender Private Key"
        class="text-truncate form-control"
        name="fromSK"
        readonly=""
        type="text"
        value="308187020100301306072a8648ce3d020106082a8648ce3d030107046d306b020101042047979df7cebe59dd7bf901"
      />
    </div>
    <small
      class="text-muted form-text"
    >
      Not shared with anyone, keep this secret!
    </small>
    <button
      aria-label="Sign Button"
      class="mt-2 btn btn-primary btn-block"
      type="submit"
    >
      <b>
        Sign
      </b>
    </button>
  </form>
</DocumentFragment>
`;
